File: Data Structure I/python/Day 8/reverse_linked_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        '''
        Approach: 
          1- Three pointer reverse, current, next. At any position in linked list they will refering to nodes like this
              O --> O --> O
              ^     ^     ^
        reverse  current next

          2- Now Assign current.next to reverse and and reverse = current and then move every pointer right 
          3- Order of these operations matters
          4- finally return reverse 
        '''
        reverse=head
        if(head==None):
            return head
        
        current=head.next
        head.next=None
        next=None
        while(current!=None):
            next=current.next
            current.next=reverse
            reverse=current
            current=next
        return reverse

File: Data Structure I/java/Day 8/ReverseLinkedList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        /**
        Approach: 
          1- Three pointer reverse, current, next. At any position in linked list they will refering to nodes like this
              O --> O --> O
              ^     ^     ^
        reverse  current next

          2- Now Assign current.next to reverse and and reverse = current and then move every pointer right 
          3- Order of these operations matters
          4- finally return reverse 
        
         */
        ListNode reverse=head;
        if(head==null){
            return head;
        }
        ListNode current=head.next;
        head.next=null;
        ListNode next;
        while(current!=null){
            next=current.next;
            current.next=reverse;
            reverse=current;
            current=next;
        }
        return reverse;
    }
}